add datastore to fetch and store comments

jsheng112 opened the pull request:
Changed from hard-coded json to using datastore. Added frontend to support submitting comments. Added backend support for storing comments and retrieving comments.

Commit:
add datastore for comments

Commit:
add name as a comment field

Commit:
adjust styling of comments

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -20,32 +20,80 @@
 public class DataServlet extends HttpServlet {
+
+  private class Comment {
+      private String comment; /** content of the comment */
```
Are the comments meant to be javadoc or regular comment? If javadoc, place before the field; if regular comment, remove one of the leading asterisks.

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -20,32 +20,80 @@
 @WebServlet("data")
 public class DataServlet extends HttpServlet {
+
+  private class Comment {
```
Since the `Comment` class does not need to access any members of the `DataServlet` class, the `Comment` class can be static.

meltsufin commented on portfolio/src/main/webapp/script.js:
```diff
@@ -154,25 +154,35 @@ function setColor() {
   fetch('data').then(response => response.json()).then((data) => {
     const commentDivElement = document.getElementById('data-container');
     commentDivElement.innerHTML = '';
     for (var i = 0; i < data.length; i++) {
```
You can also use a `data.forEach` method here instead of the the for-loop.

meltsufin commented on portfolio/src/main/webapp/script.js:
```diff
@@ -154,25 +154,35 @@ function setColor() {
- * Fetches hard-coded json from the server and adds them to the DOM.
+ * Fetch comments from the server and adds them to the DOM.
  */
 function getData() {
```
Rename to `getComments()`?

meltsufin reviewed (dismissed):
Nice work! I left a couple of minor suggestions, which are not blocking.
Before you merge, let's give @elefeint also a chance to review.
Thanks!

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -20,32 +20,80 @@
-    comments.add("comment 3");
+    // create query
+    Query query = new Query("Comment").addSort("time", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
non-actionable: in real world, you'd initialize `DatastoreService` once, upon instance initialization, and not on every request.

elefeint commented on portfolio/src/main/webapp/script.js:
```diff
@@ -154,25 +154,35 @@ function setColor() {
-      commentDivElement.appendChild(
-      createDivElement(data[i] + "\n"));
+        commentDivElement.appendChild(
+        createDivElement(data[i]));
```
nit: add an indent, so `createDivElement` does not look like a standalone statement.

elefeint reviewed (changes requested):
Sorry for the delay; minor comments, and otherwise looks great.

Commit:
fix change requests

Commit:
fix indentation issue

Commit:
adjust styling of comments

Commit:
add datastore for contact and additional comment features

contact form now connected to datastore to store email messages.
There are two new features added for comments. One is the button
that can delete all comments. The other is user can input the number
of comments they want to see.

Blog post commenting features are in the process of being added, but
is currently incomplete.

Commit:
minor formatting fixes and add comments

Commit:
add commenting feature to each blog post

Each blog post now has its own commenting features beneath it
so that user can post their comments. I made additional servlets
instead of branching from "data" because at first it was fine
but then doing switch statments got confusing so I thought it
would be best to separate them into different servers based on
their functionalities.

There is an issue where if you adjust the number of comments shown
once the page reloads it would jump to the first blog post instead
of where you were (probably can be solved with cookies in another
commit).

Delete button not functional yet.

Commit:
fix styling issues and add relevant comments

jsheng112 commented:
Sorry that there are so many new commits after addressing the changes requested. I meant to put them in a separate branch but when I actually popped them from git stash I forgot I was still in this one.

meltsufin commented on portfolio/src/main/java/com/google/sps/servlets/DeleteBlogCommentsServlet.java:
```diff
@@ -0,0 +1,69 @@
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // get each result from datastore and delete comments if that comment has the specific id
```
Can you just add a filter to the `Query` to only return entities that have id equal to the parameter?
See: https://cloud.google.com/appengine/docs/standard/java/datastore/queries
Then you wouldn't have to do this loop.

meltsufin commented on portfolio/src/main/java/com/google/sps/servlets/DeleteBlogCommentsServlet.java:
```diff
@@ -0,0 +1,69 @@
+    
+    // Send an empty JSON as the response
+    response.setContentType("application/json;");
+    String json = convertToJson(new ArrayList<String>());
```
Maybe return the number of deleted comments or something like that?

meltsufin reviewed (commented):
Not sure if it's ready for review, but I just added a couple of comments in the interim.

Commit:
fix requested changes and blog comments now fully working

Commit:
set all deletion to return number deleted

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/BlogCommentServlet.java:
```diff
@@ -0,0 +1,123 @@
+public class BlogCommentServlet extends HttpServlet {
+
+  private static class PostComment {
+      private String comment; /** content of the comment */
```
Javadoc comments go before the field, not after, so that the javadoc parser can extract them.

Commit:
complete refactoring, specific comment deletion working

Commit:
blog post comments working

meltsufin commented on portfolio/src/main/webapp/comments.html:
```diff
@@ -21,7 +21,20 @@
+                <input class="button" type="submit" />
+            </form>
+            <label id="quantity-label" for="quantity">Number of comments shown (-1 means that all comments are shown below):</label>
+            <input type="number" id="quantity" name="quantity" min="-1" value="-1" onchange="getComment();"/>
```
Not a technical issue, but from the UX point of view, I would not choose an input box with a magic -1 input that needs to be explained. To avoid the -1, you can add a checkox to toggle whether to limit results at all, and only when checkbox is selected, enable the input value control. Some web sites also use a drop-down with pre-filled values like [all, 1, 5, 10, 25, 50], or something like that.

meltsufin reviewed (dismissed):
Great work on this week's assignment Jenny!
You clearly went far beyond the requirements of the assignment and that's awesome!

I left a comment on the UX, but only if you want to further enhance the web site.

Commit:
fix -1 input issue and all deletion types works

Commit:
add emoji to normal and blog comments

meltsufin reviewed (dismissed):
LGTM

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/CommentService.java:
```diff
@@ -0,0 +1,73 @@
+        // create filter
+        Filter keyFilter = new FilterPredicate("id", FilterOperator.EQUAL, id);
+        Query query = new Query("Comment").setFilter(keyFilter);
+        System.out.println("TRYING TO DELETE ID" + id);
```
Remove debugging statement.

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/DeleteBlogCommentsServlet.java:
```diff
@@ -0,0 +1,88 @@
+  /**
+   * Converts an ArrayList instance into a JSON string using the Gson library. 
+   */
+  private String convertToJson(ArrayList<String> comments) {
```
Since `convertToJson()` method exists  in the same form in two different classes, there is opportunity to refactor it into a utility class to avoid duplication -- something like `JsonUtil`.

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java:
```diff
@@ -0,0 +1,91 @@
+
+    List<Entity> results = service.findAllComments();
+    ArrayList<Comment> comments = new ArrayList<Comment>();
+    int counter = 0;
```
In the real world, you'd be doing the filtering inside of datastore to avoid fetching large amounts of unused data:
https://cloud.google.com/appengine/docs/standard/java/datastore/retrieving-query-results#setting_limits

elefeint commented on portfolio/src/main/java/com/google/sps/servlets/CommentService.java:
```diff
@@ -0,0 +1,73 @@
+        PreparedQuery results = datastore.prepare(query);
+        List<Entity> list = results.asList(FetchOptions.Builder.withDefaults());
+        if (list.isEmpty()) {
+            System.err.println("EMPTY" + id);
```
Update logging statement with more detail -- for example, "Expected a single comment with ID %s, but no results found".

elefeint reviewed (changes requested):
Looks great; the only actionable feedback is to remove the debug log statement and update the logged warning about empty result.

You don't have to address any of the refactoring comments at all; they are just for future information.

Commit:
made mobile friendly and fix index issue

Commit:
finish refactoring and remove debugging statements

elefeint reviewed (dismissed):
Looks amazing.

Commit:
add sorting options to comments

meltsufin reviewed (approved):
Nice addition with sort!

One thing to pay attention to is opportunities to reduce code duplication. For example, you have `BlogCommentsService` and a `CommentsService`. They are very similar. You can potentially replace them with a single `CommentsService`. However, this will only be possible if you make your `Comment` entity more rich with an additional attribute that would specify whether it is for the blog or the regular comments page. Then you can filter on that `type` attribute, depending on which page is using the comments. I hope this makes sense. The less code you have, the less testing you have to do, and the easier it is to maintain.
